### PR TITLE
feat(storage): add persist form to storage

### LINKED ISSUE #24: Add persist form functionality
It'll be great to add a built-in method to persist the form value to the storage. For example:

```ts
const profileForm = new FormGroup<Profile>({
  firstName: new FormControl(''),
  lastName: new FormControl(''),
  skills: new FormArray([]),
  address: new FormGroup({
    street: new FormControl(''),
    city: new FormControl('')
  })
});

profileForm.persist(key, storage);
```

Where the `key` is the storage key, and the storage is `local` or `session`.

### PR DESCRIPTION
**Pull request by tehshin**
closes #24

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: 24

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: projects/ngneat/reactive-forms/src/lib/formGroup.ts [hunk 1]
```diff
@@ -269,4 +271,8 @@ export class FormGroup<T extends Obj = any, E extends object = any> extends NgFo
   setDisable(disable = true, opts?: ControlEventOptions) {
     disableControl(this, disable, opts);
   }
+
+  persist(key: string, storage: PersistManager<T> = new LocalStorageManager<T>()) {
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/formGroup.ts:275]: The idea is to save it automatically upon any change by subscribing to `valueChanges` with a configurable `debounceTime`

Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts:275]: Ok, got you. Should the persist function return the observable or it should it be kept within the form group? I would return it in order that it can be properly unsubscribed.

Another question: Any suggestion when to restore the value from the storage?

Comment (NetanelBasal) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts:275]: 1. Yes, it should return the subscription 
2. In the function call.

File: projects/ngneat/reactive-forms/src/lib/formGroup.ts [hunk 2]
```diff
@@ -269,4 +272,18 @@ export class FormGroup<T extends Obj = any, E extends object = any> extends NgFo
...
     disableControl(this, disable, opts);
   }
+
+  persist(
+    key: string,
+    debounceTime: number = 250,
+    manager: PersistManager<T> = new LocalStorageManager<T>()
+  ): Observable<T> {
+    const formValue = manager.getValue(key);
+    this.reset(formValue);
```
Comment (tehshin) [inline projects/ngneat/reactive-forms/src/lib/formGroup.ts:282]: How should we deal with FormArrays here? Can't really make any assumptions on how the controls inside the form arrays should be created (e.g. validators etc.)

File: projects/ngneat/reactive-forms/src/lib/formGroup.ts (hunk 1 above)
Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts]: updated the PR

File: projects/ngneat/reactive-forms/src/lib/formGroup.ts (hunk 2 above)
Comment (NetanelBasal) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts:282]: https://github.com/ngneat/forms-manager#using-formarray-controls

Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts:282]: Ah, I had something like this in mind, I'll check it out in the evening.

Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.ts]: Ok, added the FormArray factory options. Also managed to limit the config to arrays and correctly infer the value types for the factory functions.

File: projects/ngneat/reactive-forms/src/lib/control-actions.ts [hunk 3]
```diff
@@ -131,3 +139,35 @@ export function selectControlValue$<T, R>(
 ): Observable<R> {
   return (control.value$ as Observable<any>).pipe(map(mapFn), distinctUntilChanged());
 }
+
+export function persistValue$<T>(control: FormGroup<T>, key: string, options: PersistOptions<T>): Observable<T> {
+  return control.valueChanges.pipe(
+    debounceTime(options.debounceTime),
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/control-actions.ts:145]: It should default to 300ms

**Comment by NetanelBasal** [comment]
LGTM. Update the docs, please.

**Comment by tehshin** [comment]
updated the docs

**Comment by NetanelBasal** [comment]
Can you support an async storage type, please? It'll be useful for Ionic apps for example.

**Comment by tehshin** [comment]
If it can wait another week then I'll certainly try do it. Currently nowhere near a capable Dev machine. ;)

**Comment by NetanelBasal** [comment]
Sure

**Comment by tehshin** [comment]
Please re-check

File: projects/ngneat/reactive-forms/src/lib/persistManager.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { Observable } from 'rxjs';
+
+export interface PersistManager<T> {
+  setValue(key: string, data: T): void;
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/persistManager.ts:4]: `setValue` could also be async

File: projects/ngneat/reactive-forms/src/lib/utils.ts [hunk 5]
```diff
@@ -9,3 +11,19 @@ export function isFunction(x: any): x is Function {
...
+  return typeof value?.then === 'function';
+}
+
+export function wrapIntoObservable<T>(value: T | Promise<T> | Observable<T>): Observable<T> {
+  if (isObservable(value)) {
+    return value;
+  }
+
+  if (isPromise(value)) {
+    return from(Promise.resolve(value));
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/utils.ts:25]: You can just call `if(isPromise(value) | isObservable()) from(value)`

File: projects/ngneat/reactive-forms/src/lib/persistManager.ts (hunk 4 above)
Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/persistManager.ts:4]: That's true, but the concrete manager that implements `setValue` could manage the async requests internally. I am not sure what exactly the library should do with the response from the requests, besides ensuring they are run in order etc.

File: projects/ngneat/reactive-forms/src/lib/control-actions.ts [hunk 6]
```diff
@@ -122,7 +130,7 @@ export function hasErrorAndDirty<T>(control: AbstractControl<T>, error: string,
 
 export function markAllDirty<T>(control: FormArray<T> | FormGroup<T>): void {
   control.markAsDirty({ onlySelf: true });
-  (control as any)._forEachChild(control => control.markAllAsDirty());
+  (control as any)._forEachChild(control => control.markAllAsDirty?.());
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/control-actions.ts:133]: Why did you add this?

File: projects/ngneat/reactive-forms/src/lib/persistManager.ts (hunk 4 above)
Comment (NetanelBasal) [reply projects/ngneat/reactive-forms/src/lib/persistManager.ts:4]: We need to handle race conditions.

File: projects/ngneat/reactive-forms/src/lib/control-actions.ts (hunk 6 above)
Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/control-actions.ts:133]: I got an exception when using the form array factories. I'll recheck why exactly that happened.

File: projects/ngneat/reactive-forms/src/lib/persistManager.ts (hunk 4 above)
Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/persistManager.ts]: I adapted the interface, please let me know if it is ok.

File: projects/ngneat/reactive-forms/src/lib/formGroup.spec.ts [hunk 7]
```diff
@@ -285,4 +295,44 @@ describe('FormGroup', () => {
...
+        getValue: jest.fn(),
+        setValue: jest.fn()
+      };
+      control.persist('key', { debounceTime, manager: persistManager }).subscribe();
+      control.getControl('name').setValue('ewan');
+      tick(debounceTime);
+      expect(persistManager.setValue).toHaveBeenCalledWith('key', control.value);
+    }));
+
+    it.each([[person], [Promise.resolve(person)], [of(person)]])(
```
Comment (NetanelBasal) [inline projects/ngneat/reactive-forms/src/lib/formGroup.spec.ts:315]: Can you please add specs for async storage?

Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.spec.ts:315]: do you mean for `setValue`?

Comment (NetanelBasal) [reply projects/ngneat/reactive-forms/src/lib/formGroup.spec.ts:315]: I mean when the manager is async. (both getValue and setValue)

Comment (tehshin) [reply projects/ngneat/reactive-forms/src/lib/formGroup.spec.ts]: Changed the tests so that the mock gets/sets value after a delay

**Comment by tehshin** [comment]
Done. Was a pleasure to give back a little to you guys for all the libraries you guys put out that ease our development. 😄
